enhancements/etcd/etcd-size-tuning.md

```diff
@@ -0,0 +1,192 @@
+We want to be able to allow for this while minimizing the risk of them setting values that cause issues.
+
+### User Stories
+
+* As an administrator, I want to change the etcd backend size to decrease the risk of quota reached errors in very large clusters.
+* As an Openshift support, I want to walk a customer through changing the etcd size in a minimal number of steps.
+* As an etcd engineer, I want to easily add and test new sizes for performance impacts on both large and small clusters.
+
+### Goals
+
+* Add a configuration option using resource.Quantity to allow for human readable values; i.e. 16GiB.
+* Remove the parameters from the podspec rendering and replace with the value.
+* Add an API to allow admins to change the value.
+* Add blocker to disallow shrinking the size once set to a larger value.
+
+### Non-Goals
+
+* Handle consuming value changes without an etcd rollout.
+
+## Proposal
+
+Allow the customer/admin access to set the backend quota directly through a human readable value. The only restriction is that they are disallowed from setting the value lower than the current.
+
+The default value or unset value would be the current, hardcoded backend quota: 8GiB. This would allow for upgrades where this configuration value wouldn't nesscessarily be set.
+
+This would allow for more flexibility, but would also likely increase the number of support cases as the memory requirements will be higher: changing quota from 8GiB to 16GiB is not double the memory, it'll be 6x the memory since each etcd member could take up to 16GiB.
+We will need to work with the perf team to understand the memory requirements of largers quotas, and the effect of the quota on things like defragmentation.
+A downside is that with the profiles, discussed in the Alternatives section, there are a discrete, small, number of permutations to test, giving more confidence of the exact effects a given profile has on performance/memory requirements.
+This is a minor consideration as setting the quota to a arbitrarily higher value does not mean that etcd will take up that much space; it will depend on the workload.
+
+### Workflow Description
+
+1. The cluster administrator decides to change the etcd backend quota from the default (8GiB) to a larger value; e.g. 20GiB.
+2. They set the new value in the API server.
+3. An etcd redeployment will be automatically run which restarts the etcd pods which consume the new value.
+
+#### Variation [optional]
+
+1. The cluster administrator wants to set the quota from a larger value (e.g. 32GiB) to a smaller one (8GiB).
+2. They attempt to set the new value in the API server.
+3. The API server should return an error notifying them that the value is invalid - not allowed to shrink the quota.
+
+### API Extensions
+
+In operator/v1/0000_12_etcd-operator_01_config-TechPreviewNoUpgrade.crd.yaml:
+- Add BackendSizeQuota of type *resource.Quantity
+  - With default '8GiB'.
+  - With validation disallow a decrease in value.
```

Comment:
If the max is 64GiB and the minimum is 8GiB, and if CEL supports saying "this value may not shrink", then it might be more natural for the API field to have a name like `backendQuotaGiB` and be an integer.

How will an admin who changes the quota know when their change is done taking effect? Is there already a field in the status portion of the API to report the size being used, for example?

Reply:
> a name like backendQuotaGiB and be an integer.

indeed. I think that's a good compromise between readability and ease of validation.

The `resource.Quantity` would give the option for fractional values like `5.6GiB` or `5600MiB`, but I don't think it's worth going below the gigabyte sizes here. ushift as an SKU would be a concern, but it's not running the CEO to manage etcd.

Reply:
In case anyone ever wonders how such a validation would look like and before it gets lost in our internal slack:

> +kubebuilder:validation:XValidation:rule="self.CmpInt64(oldSelf)>0",message="can't decrease database size"